File: peripherie/Libraries/Util/FpMath.cpp
/*
 * FpMath.cpp
 *
 *  Created on: Oct 6, 2013
 *      Author: helios
 */

#include "FpMath.h"

FpMath::FpMath() {

}

FpMath::~FpMath() {
}

float FpMath::arctan2(int32_t y, int32_t x) {
	coeff_1 = pi / 4;
	coeff_2 = 3 * coeff_1;
	abs_y = abs(y) + 1e-10      // kludge to prevent 0/0 condition
	if (x >= 0) {
		r = (x - abs_y) / (x + abs_y);
		angle = coeff_1 - coeff_1 * r;
	} else {
		r = (x + abs_y) / (abs_y - x);
		angle = coeff_2 - coeff_1 * r;
	}
	if (y < 0)
		return (-angle);     // negate if in quad III or IV
	else
		return (angle);
}

File: client/opticopterClient/src/ch/sharpsoft/arducopter/client/math/FpMath.java
package ch.sharpsoft.arducopter.client.math;

import static java.lang.Math.abs;

public class FpMath {
	static int factor = 1 << 29;// (int) Math.pow(2, 29);
	static int coeff_1 = (int) ((Math.PI / 4) * factor);
	static int coeff_2 = 3 * coeff_1;

	/**
	 * @return in rad * 2^29
	 */
	public static int arctan2(final int y, final int x) {
		int angle;
		int abs_y = abs(y);
		if (x >= 0) {
			if (((x + abs_y) == 0)) {
				abs_y++;
			}
			int r = (int) ((long) coeff_1 * (x - abs_y)) / (x + abs_y);
			angle = coeff_1 - coeff_1 * r;
		} else {
			if (((abs_y - x) == 0)) {
				abs_y++;
			}
			int r = (int) ((long) coeff_1 * (x + abs_y)) / (abs_y - x);
			angle = coeff_2 - r;
		}
		if (y < 0)
			return (-angle); // negate if in quad III or IV
		else
			return (angle);
	}
}
